### PR TITLE
Expand C++ SDK documentation

### PR DESCRIPTION
**Pull request by mpwarres**
* Add docs/api_overview.md file
* Add doc comments to proxy_wasm_api.h and other header files
* Simplify top-level README.md and move build instructions into docs/building.md file

### CODE REVIEW

File: docs/api_overview.md [hunk 1]
```diff
@@ -0,0 +1,471 @@
...
+
+`Context` instances are associated with individual streams--i.e. incoming HTTP
+or gRPC requests, or TCP streams if the network proxy is acting as an L4
+proxy. `Context` instances receive callbacks corresponding to different events
+in the lifecycle of a stream:
+
+* `onCreate`: called when handling of a new stream starts.
+* `onDone`: called when the host is done processing the stream.
+* `onLog`: called after the host is done processing the stream, if the plugin is
+  being used for access logging.
```
Comment (martijneken) [inline docs/api_overview.md:116]: I'm not sure what this means. How does a plugin sign up for access logging? Clarify that this is not related to logging described below?

File: docs/api_overview.md [hunk 2]
```diff
@@ -0,0 +1,471 @@
...
+* `onResponseHeaders`: called when HTTP or gRPC response headers are received.
+* `onResponseBody`: called when HTTP or gRPC response body data is received.
+* `onResponseTrailers`: called when HTTP or gRPC response trailers are received.
+* `onNewConnection`: called when a new connection is established.
+* `onDownstreamData`: called when a new chunk of data is received from
+  downstream over a connection.
+* `onUpstreamData`: called when a new chunk of data is received from upstream
+  over a connection.
+
+For API details, see doc comments for the `Context` class in [proxy_wasm_api.h].
```
Comment (martijneken) [inline docs/api_overview.md:134]: You might say something here about return values and immediate responses.

**Review by martijneken — COMMENTED** [review]
Helpful docs, thank you!

File: docs/building.md [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+```bash
+docker run -v $PWD:/work -v $PWD/../envoy/api/wasm/cpp:/work/sdk -w /work  wasmsdk:v2 bash /build_wasm.sh
+```
+
+### Using Abseil from the Docker image
+
+Abseil (optionally) is built in /root/abseil and can be used. Note that the
+Abseil containers (e.g. `absl::flat_hash_set`) exercise many syscalls which are
+not supported. Consequentially individual files should be pulled in which are
+relatively self contained (e.g. `strings`). Example customized Makefile:
```
[the remark this answers is not in the document]
Comment (mpwarres) [reply docs/building.md:107]: I agree--we should determine which missing hostcalls Abseil expects. But that feels outside the scope of documentation, let's deal with that separately.

File: docs/api_overview.md [hunk 4]
```diff
@@ -0,0 +1,471 @@
...
+code instantiates a `RegisterContextFactory` in a static
+variable. `RegisterContextFactory` takes as constructor params `std::function`s
+for creating new root context and stream context instances. The plugin can use
+these to create and return instances of its own subclasses of `RootContext` and
+`Context`. For example:
+
+```c++
+static RegisterContextFactory register_ExampleContext(
+    CONTEXT_FACTORY(ExampleContext), ROOT_FACTORY(ExampleRootContext),
+    "my_root_id");
```
[the remark this answers is not in the document]
Comment (mpwarres) [reply docs/api_overview.md]: Done.

File: docs/api_overview.md (hunk 2 above)
Comment (mpwarres) [reply docs/api_overview.md:145]: Added.

File: docs/api_overview.md [hunk 5]
```diff
@@ -0,0 +1,471 @@
...
+in the lifecycle of a stream:
+
+* `onCreate`: called when handling of a new stream starts.
+* `onDone`: called when the host is done processing the stream.
+* `onLog`: called after the host is done processing the stream, if the plugin is
+  being used for access logging.
+* `onDelete`: called after the plugin has completed all processing related to
+  the stream, as indication to release resources.
+
+`Context` instances also receive callbacks corresponding to stream events:
```
[the remark this answers is not in the document]
Comment (mpwarres) [reply docs/api_overview.md]: Done.

File: docs/api_overview.md [hunk 6]
```diff
@@ -0,0 +1,471 @@
...
+prepended to the log message:
+
+* `LOG_TRACE`
+* `LOG_DEBUG`
+* `LOG_INFO`
+* `LOG_WARN`
+* `LOG_ERROR`
+* `LOG_CRITICAL`
+
+Additionally, there is a LOG macro that accepts log level as a
```
[the remark this answers is not in the document]
Comment (mpwarres) [reply docs/api_overview.md]: Removed.

File: docs/building.md [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+git apply wasm-patches/*.patch
+./autogen.sh
+emconfigure ./configure --disable-shared CXXFLAGS="-O3 -flto"
+emmake make
+cd ..
+cp protobuf-wasm/src/.libs/libprotobuf-lite.a ${CPP_API}/libprotobuf-lite.a
+cp protobuf-wasm/src/.libs/libprotobuf.a ${CPP_API}/libprotobuf.a
+```
+
+### WAVM binaries
```
[the remark this answers is not in the document]
Comment (mpwarres) [reply docs/building.md]: Removed.

**Comment by mpwarres** [comment]
Thanks for the comments, PTAL

File: docs/api_overview.md (hunk 1 above)
Comment (mpwarres) [reply docs/api_overview.md]: This is when configured as an Envoy AccessLogger (for example). Added a link to the relevant Envoy documentation.

File: docs/api_overview.md [hunk 8]
```diff
@@ -0,0 +1,471 @@
...
+## Handling TCP streams
+
+Plugins that handle TCP (or TCP-like) connections do so by overriding one or
+more of the connection-related callback methods declared by the `Context` class
+(see the [Context] section for a list). For example, a plugin that wanted to
+inspect data sent by the upstream client could override the
+`Context::onUpstreamData` method.
+
+To access the actual data being proxied, plugin code would use the
+buffer-related hostcalls described in [Buffers], specifying
```
[the remark this answers is not in the document]
Comment (mpwarres) [reply docs/api_overview.md]: Added.

File: proxy_wasm_enums.h [hunk 9]
```diff
@@ -23,23 +23,77 @@
...
+  // The host should continue to process the stream.
   Continue = 0,
+  // The host should suspend further processing of headers until plugin code
+  // unpauses the stream via a call to `continueRequest` or `continueResponse`.
   StopIteration = 1,
+  // The host should terminate the stream.
   ContinueAndEndStream = 2,
+  // The host should suspend further processing of the stream until plugin code
+  // unpauses the stream via a call to `continueRequest` or `continueResponse`,
+  // in the meantime buffering all body bytes received.
```
Comment (martijneken) [inline proxy_wasm_enums.h:52]: Kind of a critical bit of documentation from Envoy states:
```
  // If buffering the request causes buffered data to exceed the configured buffer limit, a 413 will
  // be sent to the user. On the response path exceeding buffer limits will result in a 500.
```

As a result, proxy_wasm_cpp_host is particularly opinionated about this state for headers -- should we mention that it's typically/often disallowed?
https://github.com/proxy-wasm/proxy-wasm-cpp-host/blob/611d5d94bb10aa5c30174e700d9031a33ba09d55/src/context.cc#L496-L501

See also discussion in https://github.com/proxy-wasm/proxy-wasm-cpp-host/issues/367

P.S. Hmmm FilterDataStatus (BODY processing) doesn't seem to have the same restriction...

File: README.md [hunk 10]
```diff
@@ -1,227 +1,33 @@
...
+* An [ABI](https://github.com/proxy-wasm/spec) that specifies the low-level
+  interface between network proxies and Wasm virtual machines that run the
+  plugins.
+* [Host implementations](https://github.com/proxy-wasm/spec#host-environments)
+  of the ABI, provided by network proxies.
+* [Language-specific SDKs](https://github.com/proxy-wasm/spec#sdks) that layer
+  on top of the ABI, providing a more natural and programmer-friendly API for
+  invoking and implementing Proxy-Wasm functions and callbacks.
+  
+This repository provides the C++ SDK.
```
Comment (martijneken) [inline README.md:22]: Worth mentioning here that this uses Emscripten to compile C to wasm? (and linking to https://emscripten.org/)

File: docs/api_overview.md [hunk 11]
```diff
@@ -0,0 +1,472 @@
...
+  [WasmAccessLog](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/access_loggers/wasm/v3/wasm.proto#extensions-access-loggers-wasm-v3-wasmaccesslog)
+  extension).
+* `onDelete`: called after the plugin has completed all processing related to
+  the stream, as indication to release resources.
+
+`Context` instances also receive callbacks corresponding to stream events. For
+HTTP or gRPC streams, these are:
+
+* `onRequestHeaders`: called when HTTP or gRPC request headers are received.
+* `onRequestBody`: called when HTTP or gRPC request body data is received.
```
Comment (martijneken) [inline docs/api_overview.md:125]: For the BODY handlers, say something about chunking? e.g. add "(for each chunk)"

File: docs/api_overview.md [hunk 12]
```diff
@@ -0,0 +1,472 @@
...
+passing or receiving HTTP response body data.
+
+Certain buffer types are only accessible during certain callbacks. For example,
+the `WasmBufferType::HttpResponseBody` buffer is only accessible during
+`Context::onResponseBody` and `Context::onLog` callbacks. See doc comments for
+individual callbacks in [proxy_wasm_api.h] for which buffers are accessible in
+each callback. Documentation for the `WasmData` class is also in
+[proxy_wasm_api.h].
+
+## Handling HTTP and gRPC requests
```
Comment (martijneken) [inline docs/api_overview.md:172]: This doesn't cover BODY chunks, so maybe call this "Handling HTTP headers and trailers" ?

File: docs/building.md [hunk 13]
```diff
@@ -0,0 +1,228 @@
+# Proxy-Wasm C++ SDK Build Instructions
+
+The C++ SDK has dependencies on specific versions of the C++ WebAssembly
+toolchain [Emscripten](https://emscripten.org) and the protobuf library,
+therefore use of a Docker image is recommended.
+
+## Docker
+
+A Dockerfile for the C++ SDK is provided in [Dockerfile-sdk](Dockerfile-sdk).
```
Comment (martijneken) [inline docs/building.md:9]: Broken link: it's in the parent dir

File: docs/building.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (martijneken) [reply docs/building.md]: Should we also remove this from sdk_container.sh? Not sure why it's there.

**Review by martijneken — APPROVED** [review]

File: README.md [hunk 14]
```diff
@@ -1,227 +1,33 @@
-# WebAssembly for Proxies (C++ SDK)
+# Proxy-Wasm C++ SDK
```
Comment (PiotrSikora) [inline README.md:1]: Nit: All REAMDEs use "WebAssembly for Proxies (xxx)" format, so it might make sense to leave this as is... or we make this change everywhere.

File: README.md (hunk 10 above)
Comment (PiotrSikora) [reply README.md:22]: That's an implementation detail, and honestly, it's not clear if that's something that provides any value at this point compared to using `clang --target=wasm32-wasi` (@mpw it might be worth filing an issue about that).

File: docs/api_overview.md [hunk 15]
```diff
@@ -0,0 +1,472 @@
...
+  
+  * They are the receivers of Proxy-Wasm callbacks: for example, when HTTP
+    request headers are received, that is represented as a call to the
+    `onRequestHeaders` method of a Context object. Callbacks are declared as
+    virtual methods on a `Context` base class, which plugin code can override.
+  * They provide the interface for most hostcalls, particularly those associated
+    with the current stream. The hostcalls are defined as methods on `Context`
+    base classes, which plugin code can call.
+    
+  There are two types of Contexts: *Root Contexts* and *Stream Contexts*
```
Comment (PiotrSikora) [inline docs/api_overview.md:36]: "Root Context" is a leftover from forever-ago, I'd change it to "Plugin Context (also known as Root Context)" to match the spec docs.

File: docs/api_overview.md (hunk 11 above)
Comment (PiotrSikora) [reply docs/api_overview.md:125]: +1 to document chunking, because the callback names are not suggesting that.

File: docs/api_overview.md [hunk 16]
```diff
@@ -0,0 +1,472 @@
...
+* `getRequestHeaderSize`: returns the current number of request header fields.
+
+A similar set of methods is provided for each of: request trailers, response
+headers, and response trailers.
+
+gRPC requests received by the proxy are dispatched to the plugin using the same
+`Context` callback methods as HTTP requests. Plugin code can determine whether
+an incoming request is gRPC in the same way that network proxies do: by checking
+if the ":method" pseudoheader has value "POST" and the "content-type" header has
+value "application/grpc".
```
Comment (PiotrSikora) [inline docs/api_overview.md:211]: `s/has/starts with/`

File: proxy_wasm_enums.h (hunk 9 above)
Comment (PiotrSikora) [reply proxy_wasm_enums.h:52]: If I recall correctly, the issue we had was around pausing headers, while processing body, so this restriction was added only for headers and not to other return values.

File: docs/api_overview.md [hunk 17]
```diff
@@ -0,0 +1,472 @@
...
+* [proxy_wasm_enums.h](../proxy_wasm_enums.h): supporting enums for the API
+* [proxy_wasm_intrinsics.js](../proxy_wasm_intrinsics.js): list of Proxy-Wasm
+  ABI hostcalls, for use by [Emscripten](https://emscripten.org)
+* [proxy_wasm_intrinsics.proto](../proxy_wasm_intrinsics.proto): protobuf types
+  needed for gRPC calls
+* [proxy_wasm_intrinsics.h](../proxy_wasm_intrinsics.h): combined header file
+  that includes all other header files
+* [proxy_wasm_intrinsics.cc](../proxy_wasm_intrinsics.cc): implementation of
+  dispatch from ABI-level Proxy-Wasm callbacks to [Context] and [RootContext]
+  callback methods.
```
Comment (PiotrSikora) [inline docs/api_overview.md:462]: Nit: Maybe move all link targets to the bottom of this file for consistency?

File: docs/building.md (hunk 13 above)
Comment (PiotrSikora) [reply docs/building.md:9]: +1

File: docs/building.md [hunk 18]
```diff
@@ -0,0 +1,228 @@
...
+
+It can built in this directory by:
```
Comment (PiotrSikora) [inline docs/building.md:11]: in this repository's root directory

File: docs/building.md [hunk 19]
```diff
@@ -0,0 +1,228 @@
...
+
+This will save time on subsequent compiles.
+
+### Using the SDK from a newer/specific version of Envoy
+
+To use a newer/specific version of the SDK (e.g. from the version of Envoy you
+are going to deploy the WebAssembly module to), bind that volume and use it in
+the Makefile.
+
+Here is an example Makefile referencing the SDK at `../envoy/api/wasm/cpp` and
```
Comment (PiotrSikora) [inline docs/building.md:84]: This is leftover from when C++ SDK was hosted inside Envoy's codebase.

File: docs/building.md [hunk 20]
```diff
@@ -0,0 +1,228 @@
...
+```
+
+Run docker pointing to Envoy sources in a directory parallel (at the same level)
+as your project directory:
+
+```bash
+docker run -v $PWD:/work -v $PWD/../envoy/api/wasm/cpp:/work/sdk -w /work  wasmsdk:v2 bash /build_wasm.sh
+```
+
+### Using Abseil from the Docker image
```
Comment (PiotrSikora) [inline docs/building.md:102]: I don't think we're using Abseil in C++ SDK anymore.

File: docs/building.md [hunk 21]
```diff
@@ -0,0 +1,228 @@
...
+./emsdk activate latest
+```
+
+However 3.1.7 is known to work.
+
+### Rebuilding the libprotobuf.a files
+
+If want to rebuild the libprotobuf.a files or use a different version see the
+instructions at https://github.com/kwonoj/protobuf-wasm. Commit
+4bba8b2f38b5004f87489642b6ca4525ae72fe7f works for protobuf v3.9.x.
```
Comment (PiotrSikora) [inline docs/building.md:211]: Do we still need this hack and documentation for protobufs?

It looks that upstream added support for building with Emscripten: https://github.com/protocolbuffers/protobuf/commit/8e869f76275253726da5a712de4d4a1b4de22ca2

File: proxy_wasm_enums.h (hunk 9 above)
Comment (mpwarres) [reply proxy_wasm_enums.h]: Added text to `StopIteration` noting that some hosts may handle it the same as `StopAllIterationAndWatermark`. Also added text mentioning that any buffering is subject to host limits.

File: README.md (hunk 14 above)
Comment (mpwarres) [reply README.md]: Done.

File: README.md (hunk 10 above)
Comment (mpwarres) [reply README.md:22]: The previous language and link for Emscripten wasn't removed, it was just moved to the building.md page, since it's more of a build/implementation detail.

Filed #167 for updating SDK to document using `clang --target=wasm32-wasi` directly.

File: docs/building.md (hunk 13 above)
Comment (mpwarres) [reply docs/building.md]: Done

File: docs/building.md (hunk 18 above)
Comment (mpwarres) [reply docs/building.md]: Done

File: docs/building.md (hunk 19 above)
Comment (mpwarres) [reply docs/building.md]: Thanks, removed

File: docs/building.md (hunk 20 above)
Comment (mpwarres) [reply docs/building.md:78]: Currently, sdk_container.sh still bundles a version of Abseil into the Docker image. Rather than change any behavior in this PR, I'd prefer to keep mention of Abseil until we remove that.

File: docs/building.md (hunk 21 above)
Comment (mpwarres) [reply docs/building.md:187]: Most likely not. I've kept the instructions in place for now, until #159 is fixed, but with an added note that the hack should only be needed for protobuf versions earlier than v3.15.

File: docs/api_overview.md (hunk 11 above)
Comment (mpwarres) [reply docs/api_overview.md]: Done

File: docs/api_overview.md (hunk 12 above)
Comment (mpwarres) [reply docs/api_overview.md]: Done

File: docs/api_overview.md (hunk 16 above)
Comment (mpwarres) [reply docs/api_overview.md]: Done

File: docs/api_overview.md (hunk 17 above)
Comment (mpwarres) [reply docs/api_overview.md]: Done

**Comment by mpwarres** [comment]
@PiotrSikora this is ready for another look. Thanks!

**Review by PiotrSikora — APPROVED** [review]
Thanks!
